File: src/client/plain.ts
import * as readline from 'readline';
import chalk from 'chalk';
import type { MinimapCell, OnlinePlayer, OutputStyle, PlayerSnapshot } from '../protocol/messages.js';

const STYLES: Record<OutputStyle, (s: string) => string> = {
  normal: (s) => s,
  system: chalk.cyan,
  combat: chalk.red,
  chat: chalk.yellow,
  quest: chalk.magenta,
  loot: chalk.green,
  death: chalk.red.bold,
  party: chalk.blue,
  trade: chalk.hex('#FFA500'),
  global: chalk.magenta,
  emote: chalk.cyan.italic,
  epic: chalk.magenta.bold,
};

export interface RoomView {
  title: string;
  description: string;
  exits: string;
  entities: string[];
  zone?: string;
  minimap?: MinimapCell[];
  zoneArt?: string;
}

export interface GameClientUI {
  onInput(handler: (line: string) => void): void;
  showBanner(): void;
  log(text: string, style?: OutputStyle): void;
  showRoom(room: RoomView): void;
  showStats(player: PlayerSnapshot): void;
  showOnline(players: OnlinePlayer[]): void;
  showTicker(text: string): void;
  showMotd(text: string): void;
  flash(color: 'red' | 'yellow' | 'green'): void;
  bell(): void;
  showClassSelect(): void;
  showError(text: string): void;
  setPrompt(text: string, hidden?: boolean): void;
  showDisconnect(reason: string): void;
  destroy(): void;
}

export function createPlainClient(): GameClientUI {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout, terminal: true });
  let inputHandler: ((line: string) => void) | null = null;

  rl.on('line', (line) => inputHandler?.(line));

  return {
    onInput(handler) {
      inputHandler = handler;
    },

    showBanner() {
      console.log(chalk.bold.cyan(`
╔══════════════════════════════════════════════════════════╗
║              R E A L M   O F   E C H O E S              ║
║         A Classic MMO Text Adventure                     ║
╚══════════════════════════════════════════════════════════╝
`));
    },

    log(text, style = 'normal') {
      const formatter = STYLES[style] ?? STYLES.normal;
      for (const line of text.split('\n')) {
        console.log(formatter(line));
      }
    },

    showRoom(room) {
      console.log();
      if (room.zoneArt) console.log(chalk.dim(room.zoneArt));
      console.log(chalk.bold.white(`[ ${room.title} ]`));
      console.log(chalk.white(room.description));
      if (room.exits) console.log(chalk.dim(`Exits: ${room.exits}`));
      if (room.entities.length > 0) {
        console.log(chalk.dim('Also here:'));
        for (const e of room.entities) {
          const hostile = e.includes('[hostile]');
          console.log(chalk.dim('  ') + (hostile ? chalk.red(`• ${e}`) : chalk.green(`• ${e}`)));
        }
      }
      console.log();
    },

    showStats(player) {
      const hp = bar(player.hp, player.maxHp, chalk.red, chalk.dim.red);
      const mp = bar(player.mp, player.maxMp, chalk.blue, chalk.dim.blue);
      const duel = player.inDuel ? chalk.red(' [DUEL]') : '';
      console.log(
        chalk.dim(` Lv.${player.level} `) + hp + chalk.dim(` ${player.hp}/${player.maxHp} `) +
        mp + chalk.dim(` ${player.mp}/${player.maxMp} `) +
        chalk.yellow(`${player.gold}g`) + duel,
      );
    },

    showOnline(players) {
      console.log(chalk.dim(`-- ${players.length} online --`));
      for (const p of players) {
        console.log(chalk.dim(`  ${p.username} Lv.${p.level} ${p.className} @ ${p.zone}`));
      }
    },

    showTicker(text) {
      console.log(chalk.dim(`› ${text}`));
    },

    showMotd(text) {
      console.log(chalk.cyan(`\n=== MOTD ===\n${text}\n`));
    },

    flash() {},

    bell() {
      process.stdout.write('\x07');
    },

    showClassSelect() {
      console.log(chalk.cyan('\nChoose class: warrior | mage | rogue\n'));
    },

    showError(text) {
      console.log(chalk.red(`Error: ${text}`));
    },

    setPrompt(text) {
      rl.setPrompt(text);
      rl.prompt();
    },

    showDisconnect(reason) {
      console.log(chalk.dim(reason));
    },

    destroy() {
      rl.close();
    },
  };
}

function bar(
  current: number,
  max: number,
  fill: (s: string) => string,
  empty: (s: string) => string,
): string {
  const width = 8;
  const filled = Math.round((current / max) * width);
  return fill('█'.repeat(filled)) + empty('░'.repeat(Math.max(0, width - filled)));
}

File: crates/realm-client/src/plain.rs
use std::io::{self, Write};

use anyhow::Result;
use realm_protocol::{OnlinePlayer, OutputStyle, PlayerSnapshot, ServerMessage};
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::sync::mpsc;

use crate::app::{
    handle_user_input, reconnect_delay, reconnect_limit_reached, run_connection, should_reconnect,
    ClientState, WsEvent,
};

pub struct RoomView {
    pub title: String,
    pub description: String,
    pub exits: String,
    pub entities: Vec<String>,
    pub zone_art: Option<String>,
}

pub async fn run(server_url: &str) -> Result<()> {
    let mut state = ClientState::default();
    let mut stdin = BufReader::new(tokio::io::stdin());
    let mut line = String::new();

    loop {
        let (ws_tx, mut ws_rx) = mpsc::unbounded_channel::<WsEvent>();
        let (cmd_tx, cmd_rx) = mpsc::unbounded_channel();
        let url = server_url.to_string();
        let conn_state = state.clone();
        let conn_task = tokio::spawn(async move {
            let _ = run_connection(&url, &conn_state, ws_tx, cmd_rx).await;
        });

        let mut connected = false;

        'session: loop {
            tokio::select! {
                event = ws_rx.recv() => {
                    let Some(event) = event else { break 'session };
                    match event {
                        WsEvent::Connected => {
                            connected = true;
                            state.reconnect_attempts = 0;
                        }
                        WsEvent::Message(msg) => {
                            if let Some(prompt) = apply_message(&mut state, msg) {
                                print!("{prompt}");
                                io::stdout().flush().ok();
                            }
                        }
                        WsEvent::Error(err) => {
                            if !connected && state.reconnect_attempts == 0 && !state.authenticated {
                                eprintln!("Connection error: {err}");
                                eprintln!("Is the server running? Start it with: cargo run -p realm-server");
                                conn_task.abort();
                                return Ok(());
                            }
                        }
                        WsEvent::Disconnected => {
                            if state.intentional_disconnect {
                                println!("Farewell, adventurer!");
                                conn_task.abort();
                                return Ok(());
                            }
                            if should_reconnect(&state) {
                                state.reconnect_attempts += 1;
                                println!(
                                    "Connection lost. Reconnecting ({}/{})...",
                                    state.reconnect_attempts, 5
                                );
                                break 'session;
                            }
                            if reconnect_limit_reached(&state) || !state.authenticated {
                                println!("Disconnected from server.");
                                conn_task.abort();
                                return Ok(());
                            }
                        }
                    }
                }
                read = stdin.read_line(&mut line) => {
                    if read? > 0 {
                        let input = line.trim_end().to_string();
                        line.clear();
                        if let Some(msg) = handle_user_input(&mut state, &input) {
                            let _ = cmd_tx.send(msg);
                        } else if !state.authenticated && state.auth_step == crate::app::AuthStep::Mode {
                            // auth helper messages
                            match input.to_lowercase().as_str() {
                                "login" => println!("Enter your username."),
                                "register" => println!("Choose a username (3-16 chars)."),
                                _ if !input.is_empty() => eprintln!("Error: Type \"login\" or \"register\"."),
                                _ => {}
                            }
                        } else if state.auth_step == crate::app::AuthStep::Password && state.pending_password != "__register__" {
                            println!("Enter password.");
                        } else if state.auth_step == crate::app::AuthStep::Class {
                            if !input.is_empty() {
                                eprintln!("Error: Choose: warrior, mage, or rogue");
                            }
                            show_class_select();
                        }
                        print!("{}", state.prompt);
                        io::stdout().flush().ok();
                    } else {
                        conn_task.abort();
                        return Ok(());
                    }
                }
            }
        }

        conn_task.abort();
        reconnect_delay(state.reconnect_attempts).await;
    }
}

fn apply_message(state: &mut ClientState, msg: ServerMessage) -> Option<String> {
    match msg {
        ServerMessage::Banner => {
            show_banner();
            None
        }
        ServerMessage::Output { text, style } => {
            log(&text, style.unwrap_or(OutputStyle::Normal));
            None
        }
        ServerMessage::Room {
            title,
            description,
            exits,
            entities,
            zone_art,
            ..
        } => {
            show_room(&RoomView {
                title,
                description,
                exits,
                entities,
                zone_art,
            });
            None
        }
        ServerMessage::Stats { player } => {
            show_stats(&player);
            None
        }
        ServerMessage::Online { players } => {
            show_online(&players);
            None
        }
        ServerMessage::Flash { .. } => None,
        ServerMessage::Bell => {
            print!("\x07");
            io::stdout().flush().ok();
            None
        }
        ServerMessage::Ticker { text } => {
            println!("› {text}");
            None
        }
        ServerMessage::Motd { text } => {
            println!("\n=== MOTD ===\n{text}\n");
            None
        }
        ServerMessage::Error { text } => {
            eprintln!("Error: {text}");
            None
        }
        ServerMessage::Prompt { text } => {
            if state.authenticated {
                state.prompt = ">".into();
            } else {
                state.prompt = text.trim().to_string();
                if state.prompt.is_empty() {
                    state.prompt = "login or register?".into();
                }
            }
            Some(format!("{} ", state.prompt))
        }
        ServerMessage::Disconnect { reason } => {
            state.intentional_disconnect = true;
            println!("{reason}");
            None
        }
        ServerMessage::Combat { .. } => None,
    }
}

fn show_banner() {
    println!(
        r#"
╔══════════════════════════════════════════════════════════╗
║              R E A L M   O F   E C H O E S              ║
║         A Classic MMO Text Adventure                     ║
╚══════════════════════════════════════════════════════════╝
"#
    );
}

fn log(text: &str, style: OutputStyle) {
    let prefix = match style {
        OutputStyle::Combat => "⚔ ",
        OutputStyle::System => "» ",
        _ => "",
    };
    for line in text.lines() {
        println!("{prefix}{line}");
    }
}

fn show_room(room: &RoomView) {
    println!();
    if let Some(art) = &room.zone_art {
        println!("{art}");
    }
    println!("[ {} ]", room.title);
    println!("{}", room.description);
    if !room.exits.is_empty() {
        println!("Exits: {}", room.exits);
    }
    if !room.entities.is_empty() {
        println!("Also here:");
        for entity in &room.entities {
            println!("  • {entity}");
        }
    }
    println!();
}

fn show_stats(player: &PlayerSnapshot) {
    let duel = if player.in_duel.unwrap_or(false) { " [DUEL]" } else { "" };
    println!(
        " Lv.{} HP {}/{} MP {}/{} {}g{}",
        player.level, player.hp, player.max_hp, player.mp, player.max_mp, player.gold, duel
    );
}

fn show_online(players: &[OnlinePlayer]) {
    println!("-- {} online --", players.len());
    for p in players {
        println!(
            "  {} Lv.{} {} @ {}",
            p.username,
            p.level,
            p.class_name.as_str(),
            p.zone
        );
    }
}

fn show_class_select() {
    println!("\nChoose class: warrior | mage | rogue\n");
}